File: 257.binary-tree-paths.cpp
/*
 * @lc app=leetcode id=257 lang=cpp
 *
 * [257] Binary Tree Paths
 */

#include <string>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// @lc code=start
/**
 * Definition for a binary tree node.
 */
class Solution
{
public:
    vector<string> binaryTreePaths(TreeNode* root)
    {
        vector<string> res;
        string path;
        dfs(root, path, res);
        return res;
    }

private:
    void dfs(TreeNode *n, string &path, vector<string> &res)
    {
        if (!n) {
            return;
        }
        if (!n->left && !n->right) {
            res.push_back(path + to_string(n->val));
            return;
        }

        auto s = path.size();
        path += to_string(n->val) + "->";
        dfs(n->left, path, res);
        dfs(n->right, path, res);
        path.resize(s);
    }
};
// @lc code=end

File: 257.binary-tree-paths.py
#
# @lc app=leetcode id=257 lang=python3
#
# [257] Binary Tree Paths
#

from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        res = []

        def dfs(node: Optional[TreeNode], path: str) -> None:
            if not node:
                return

            if not node.left and not node.right:
                res.append(path + str(node.val))
                return

            path += f"{node.val}->"

            dfs(node.left, path)
            dfs(node.right, path)

        dfs(root, "")
        return res

        
# @lc code=end

